File: source/c/struct_of_arrays.c
// gcc -o struct_of_arrays struct_of_arrays.c -lm
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define COUNT 5

double *xs, *ys, *zs;

void set_point(int i, double x, double y, double z) {
    xs[i] = x; ys[i] = y; zs[i] = z;
}

double distance(int i) {
    return sqrt(xs[i]*xs[i] + ys[i]*ys[i] + zs[i]*zs[i]);
}

int main(void) {
    xs = malloc(COUNT * sizeof(double));
    ys = malloc(COUNT * sizeof(double));
    zs = malloc(COUNT * sizeof(double));

    set_point(0, 1.0, 0.0, 0.0);
    set_point(1, 0.0, 2.0, 0.0);
    set_point(2, 0.0, 0.0, 3.0);
    set_point(3, 1.0, 1.0, 1.0);
    set_point(4, 3.0, 4.0, 0.0);

    for (int i = 0; i < COUNT; i++) {
        double d = distance(i);
        // Print with same format as Hemlock (removes trailing zeros)
        // Hemlock prints 1 as 1, 1.7320508... as full float
        // Use %g-like formatting
        printf("point %d: (%.1f, %.1f, %.1f) distance = %g\n",
               i, xs[i], ys[i], zs[i], d);
    }

    free(xs); free(ys); free(zs);
    printf("all arrays freed\n");
    return 0;
}

File: source/python/struct_of_arrays.py
import math

COUNT = 5
xs = [0.0] * COUNT
ys = [0.0] * COUNT
zs = [0.0] * COUNT

def set_point(i, x, y, z):
    xs[i] = x
    ys[i] = y
    zs[i] = z

def distance(i):
    x, y, z = xs[i], ys[i], zs[i]
    return math.sqrt(x * x + y * y + z * z)

set_point(0, 1.0, 0.0, 0.0)
set_point(1, 0.0, 2.0, 0.0)
set_point(2, 0.0, 0.0, 3.0)
set_point(3, 1.0, 1.0, 1.0)
set_point(4, 3.0, 4.0, 0.0)

for i in range(COUNT):
    x, y, z = xs[i], ys[i], zs[i]
    d = distance(i)
    # Format floats to match Hemlock output (remove trailing zeros where possible)
    def fmt(v):
        if v == int(v):
            return str(int(v)) + ".0"
        return str(v)
    print(f"point {i}: ({fmt(x)}, {fmt(y)}, {fmt(z)}) distance = {d}")

print("all arrays freed")
